File: bridge/internal/tui/tui.go
package tui

import (
	"fmt"
	"os"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/charmbracelet/x/ansi"
	"github.com/kartben/zephyr-in-the-browser/bridge/internal/server"
	"golang.org/x/term"
)

// Host is the subset of *server.Bridge the TUI drives.
type Host interface {
	Snapshot() server.Snapshot
	SubscribeStatus(func(server.Snapshot)) (unsubscribe func())
	RefreshPorts() error
	StopSerial()
	SelectSerial(path string, baud int) error
	AttachGdb(host string, port int) error
	DetachGdb()
	Close() error
}

type statusMsg server.Snapshot

type model struct {
	host     Host
	baud     int
	snap     server.Snapshot
	selected int
	help     bool
	errMsg   string
	width    int
}

var (
	boldStyle   = lipgloss.NewStyle().Bold(true)
	dimStyle    = lipgloss.NewStyle().Faint(true)
	cyanStyle   = lipgloss.NewStyle().Foreground(lipgloss.Color("6"))
	greenStyle  = lipgloss.NewStyle().Foreground(lipgloss.Color("2"))
	yellowStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("3"))
	redStyle    = lipgloss.NewStyle().Foreground(lipgloss.Color("1"))
)

// Run blocks until quit. Falls back to headless if stdin is not a TTY.
func Run(bridge Host, baud int) error {
	in := os.Stdin
	out := os.Stdout
	if !term.IsTerminal(int(in.Fd())) || !term.IsTerminal(int(out.Fd())) {
		return runHeadless(bridge)
	}

	m := model{
		host:  bridge,
		baud:  baud,
		snap:  bridge.Snapshot(),
		width: 80,
	}
	p := tea.NewProgram(m, tea.WithAltScreen(), tea.WithInput(in), tea.WithOutput(out))

	// Subscribe off the main goroutine: the initial status callback uses
	// Program.Send, which blocks until Run() is reading the message channel.
	subDone := make(chan struct{})
	go func() {
		unsub := bridge.SubscribeStatus(func(s server.Snapshot) {
			p.Send(statusMsg(s))
		})
		<-subDone
		unsub()
	}()

	_, err := p.Run()
	close(subDone)
	_ = bridge.Close()
	return err
}

func (m model) Init() tea.Cmd { return nil }

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		if msg.Width > 0 {
			m.width = msg.Width
		}
	case statusMsg:
		m.snap = server.Snapshot(msg)
		if m.selected >= len(m.snap.Ports) {
			m.selected = max(0, len(m.snap.Ports)-1)
		}
	case tea.KeyMsg:
		switch msg.String() {
		case "q", "ctrl+c":
			return m, tea.Quit
		case "?", "h":
			m.help = !m.help
		case "r":
			_ = m.host.RefreshPorts()
		case "s":
			m.host.StopSerial()
			m.errMsg = ""
		case "g":
			st := m.host.Snapshot().GDB
			if err := m.host.AttachGdb(st.Host, st.Port); err != nil {
				m.errMsg = err.Error()
			} else {
				m.errMsg = ""
			}
		case "G":
			m.host.DetachGdb()
		case "up", "k":
			if m.selected > 0 {
				m.selected--
			}
		case "down", "j":
			if m.selected < len(m.snap.Ports)-1 {
				m.selected++
			}
		case "enter":
			if m.selected < 0 || m.selected >= len(m.snap.Ports) {
				m.errMsg = "No port selected"
				break
			}
			m.errMsg = ""
			_ = m.host.SelectSerial(m.snap.Ports[m.selected].Path, m.baud)
		}
	}
	return m, nil
}

func (m model) View() string {
	return render(m.snap, m.selected, m.help, m.errMsg, m.width)
}

func render(s server.Snapshot, selected int, help bool, errMsg string, width int) string {
	if width <= 0 {
		width = 80
	}
	var lines []string
	add := func(line string) {
		lines = append(lines, ansi.Truncate(line, width, "…"))
	}

	add(boldStyle.Render("Zephyr bridge") + "  " + dimStyle.Render("v"+s.Version))
	add("")
	add(cyanStyle.Render("WebSocket") + "  " + s.WSURL)
	add(cyanStyle.Render("Open") + "       " + s.DeepLink)
	add("")
	add(boldStyle.Render("Serial ports") + "  " + dimStyle.Render("(↑↓ select · enter stream · s stop · r refresh)"))

	if len(s.Ports) == 0 {
		add("  " + dimStyle.Render("No serial ports yet. Plug in a board or probe."))
	} else {
		for i, p := range s.Ports {
			active := s.Serial.Path != nil && *s.Serial.Path == p.Path && s.Serial.Phase == "streaming"
			cursor := " "
			if i == selected {
				cursor = "›"
			}
			mark := dimStyle.Render("○")
			if active {
				mark = greenStyle.Render("●")
			}
			add(fmt.Sprintf("  %s %s %s  %s", cursor, mark, p.Path, dimStyle.Render(p.FriendlyName)))
		}
	}

	add("")
	add(fmt.Sprintf("%s   %s%s  %s",
		boldStyle.Render("CTF"),
		phaseStyle(s.Serial.Phase).Render(s.Serial.Phase),
		serialSuffix(s),
		dimStyle.Render(fmt.Sprintf("%d B", s.CtfBytes)),
	))
	add(fmt.Sprintf("%s   %s  %s:%d%s  %s",
		boldStyle.Render("GDB"),
		phaseStyle(s.GDB.Phase).Render(s.GDB.Phase),
		s.GDB.Host, s.GDB.Port,
		gdbDetail(s),
		dimStyle.Render(fmt.Sprintf("%d B", s.GdbBytes)),
	))

	netLabel := "unavailable"
	netStyle := dimStyle
	if s.NetReady {
		netLabel = "ready (gvisor)"
		netStyle = greenStyle
	}
	add(boldStyle.Render("Net") + "   " + netStyle.Render(netLabel))
	add(fmt.Sprintf("%s  %d/%d", boldStyle.Render("Clients"), s.Clients, s.MaxClients))
	if errMsg != "" {
		add(redStyle.Render(errMsg))
	}
	add("")
	if help {
		add(yellowStyle.Render("Keys"))
		add("  enter  stream CTF from the selected port")
		add("  s      stop serial stream")
		add("  g      attach GDB proxy (OpenOCD / J-Link on host:port)")
		add("  G      detach GDB")
		add("  r      refresh port list")
		add("  ?      toggle this help")
		add("  q      quit (Ctrl-C also works)")
		add("")
	}
	add(dimStyle.Render("[?] help  [g] GDB  [q] quit · daemon stays up if you only unplug the board"))

	return strings.Join(lines, "\n")
}

func phaseStyle(phase string) lipgloss.Style {
	switch phase {
	case "streaming", "connected":
		return greenStyle
	case "error":
		return redStyle
	default:
		return dimStyle
	}
}

func serialSuffix(s server.Snapshot) string {
	out := ""
	if s.Serial.Path != nil {
		out += fmt.Sprintf("  %s @ %d", *s.Serial.Path, s.Serial.BaudRate)
	}
	if s.Serial.Detail != "" {
		out += "  " + dimStyle.Render(s.Serial.Detail)
	}
	return out
}

func gdbDetail(s server.Snapshot) string {
	if s.GDB.Detail == "" {
		return ""
	}
	return "  " + dimStyle.Render(s.GDB.Detail)
}

func runHeadless(bridge Host) error {
	unsub := bridge.SubscribeStatus(func(s server.Snapshot) {
		fmt.Fprintf(os.Stderr, "[bridge] status clients=%d/%d serial=%s net=%v ctf=%d B\n",
			s.Clients, s.MaxClients, s.Serial.Phase, s.NetReady, s.CtfBytes)
	})
	defer unsub()
	select {}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: probe/tui.mjs
/**
 * Compact full-screen TUI for the probe bridge.
 *
 * No extra deps: alternate screen + raw stdin. Survives serial unplug; the
 * process stays up until q / Ctrl-C.
 */

import { fmtBytes } from './server.mjs'

const ESC = '\x1b'
const HIDE = `${ESC}[?25l`
const SHOW = `${ESC}[?25h`
const ALT_ON = `${ESC}[?1049h`
const ALT_OFF = `${ESC}[?1049l`
const CLEAR = `${ESC}[2J${ESC}[H`
const DIM = `${ESC}[2m`
const RESET = `${ESC}[0m`
const BOLD = `${ESC}[1m`
const GREEN = `${ESC}[32m`
const YELLOW = `${ESC}[33m`
const RED = `${ESC}[31m`
const CYAN = `${ESC}[36m`

/**
 * @param {{
 *   bridge: Awaited<ReturnType<import('./server.mjs').startProbeBridge>>,
 *   wsUrl: string,
 *   deepLink: string,
 *   baudRate: number,
 *   input?: NodeJS.ReadStream,
 *   output?: NodeJS.WriteStream,
 * }} opts
 */
export async function runTui(opts) {
  const {
    bridge,
    wsUrl,
    deepLink,
    baudRate,
    input = process.stdin,
    output = process.stdout,
  } = opts

  if (!input.isTTY || !output.isTTY) {
    return runHeadless(bridge)
  }

  let selected = 0
  let help = false
  let lastErr = ''
  /** @type {ReturnType<typeof bridge.snapshot>} */
  let snap = bridge.snapshot()

  const unsub = bridge.subscribeStatus((s) => {
    snap = s
    // Keep selection in range as ports hot-plug.
    if (selected >= snap.ports.length) selected = Math.max(0, snap.ports.length - 1)
    draw()
  })

  const draw = () => {
    const lines = []
    lines.push(`${BOLD}Zephyr probe bridge${RESET}  ${DIM}v${snap.version}${RESET}`)
    lines.push('')
    lines.push(`${CYAN}WebSocket${RESET}  ${wsUrl}`)
    lines.push(`${CYAN}Open${RESET}       ${deepLink}`)
    lines.push('')
    lines.push(
      `${BOLD}Serial ports${RESET}  ${DIM}(↑↓ select · enter stream · s stop · r refresh)${RESET}`,
    )

    if (snap.ports.length === 0) {
      lines.push(`  ${DIM}No serial ports yet. Plug in a board or probe.${RESET}`)
    } else {
      for (let i = 0; i < snap.ports.length; i++) {
        const p = snap.ports[i]
        const active = snap.serial.path === p.path && snap.serial.phase === 'streaming'
        const cursor = i === selected ? '›' : ' '
        const mark = active ? `${GREEN}●${RESET}` : `${DIM}○${RESET}`
        const label = `${p.path}  ${DIM}${p.friendlyName}${RESET}`
        lines.push(`  ${cursor} ${mark} ${label}`)
      }
    }

    lines.push('')
    const phaseColor =
      snap.serial.phase === 'streaming' ? GREEN : snap.serial.phase === 'error' ? RED : DIM
    lines.push(
      `${BOLD}CTF${RESET}   ${phaseColor}${snap.serial.phase}${RESET}` +
        (snap.serial.path ? `  ${snap.serial.path} @ ${snap.serial.baudRate}` : '') +
        (snap.serial.detail ? `  ${DIM}${snap.serial.detail}${RESET}` : '') +
        `  ${DIM}${fmtBytes(snap.ctfBytes)}${RESET}`,
    )
    const gdbColor =
      snap.gdb.phase === 'connected' ? GREEN : snap.gdb.phase === 'error' ? RED : DIM
    lines.push(
      `${BOLD}GDB${RESET}   ${gdbColor}${snap.gdb.phase}${RESET}  ${snap.gdb.host}:${snap.gdb.port}` +
        (snap.gdb.detail ? `  ${DIM}${snap.gdb.detail}${RESET}` : '') +
        `  ${DIM}${fmtBytes(snap.gdbBytes)}${RESET}`,
    )
    lines.push(
      `${BOLD}Clients${RESET}  ${snap.clients}/${snap.maxClients}`,
    )
    if (lastErr) lines.push(`${RED}${lastErr}${RESET}`)
    lines.push('')
    if (help) {
      lines.push(`${YELLOW}Keys${RESET}`)
      lines.push('  enter  stream CTF from the selected port')
      lines.push('  s      stop serial stream')
      lines.push('  g      attach GDB proxy (OpenOCD / J-Link on host:port)')
      lines.push('  G      detach GDB')
      lines.push('  r      refresh port list')
      lines.push('  ?      toggle this help')
      lines.push('  q      quit (Ctrl-C also works)')
      lines.push('')
    }
    lines.push(
      `${DIM}[?] help  [g] GDB  [q] quit · daemon stays up if you only unplug the board${RESET}`,
    )

    output.write(CLEAR + HIDE + lines.join('\n') + '\n')
  }

  const onKey = (buf) => {
    const s = buf.toString('utf8')
    if (s === '\u0003' || s === 'q' || s === 'Q') {
      cleanup()
      bridge.close().then(() => process.exit(0))
      return
    }
    if (s === '?' || s === 'h') {
      help = !help
      draw()
      return
    }
    if (s === 'r' || s === 'R') {
      void bridge.refreshPorts()
      return
    }
    if (s === 's' || s === 'S') {
      bridge.stopSerial()
      lastErr = ''
      return
    }
    if (s === 'g') {
      try {
        bridge.attachGdb()
        lastErr = ''
      } catch (err) {
        lastErr = err instanceof Error ? err.message : String(err)
        draw()
      }
      return
    }
    if (s === 'G') {
      bridge.detachGdb()
      return
    }
    // arrows: ESC [ A/B
    if (s === '\u001b[A' || s === 'k') {
      selected = Math.max(0, selected - 1)
      draw()
      return
    }
    if (s === '\u001b[B' || s === 'j') {
      selected = Math.min(Math.max(0, snap.ports.length - 1), selected + 1)
      draw()
      return
    }
    if (s === '\r' || s === '\n') {
      const p = snap.ports[selected]
      if (!p) {
        lastErr = 'No port selected'
        draw()
        return
      }
      lastErr = ''
      bridge.selectSerial(p.path, baudRate)
    }
  }

  const cleanup = () => {
    unsub()
    input.off('data', onKey)
    if (input.isTTY) {
      try {
        input.setRawMode(false)
      } catch {
        /* */
      }
    }
    input.pause()
    output.write(SHOW + ALT_OFF)
  }

  output.write(ALT_ON + HIDE)
  input.setRawMode(true)
  input.resume()
  input.on('data', onKey)
  draw()

  await new Promise(() => {
    /* runs until quit */
  })
}

async function runHeadless(bridge) {
  bridge.subscribeStatus((snap) => {
    const ser =
      snap.serial.phase === 'streaming'
        ? `${snap.serial.path} streaming ${fmtBytes(snap.ctfBytes)}`
        : snap.serial.phase
    console.log(
      `[probe] status clients=${snap.clients} serial=${ser} gdb=${snap.gdb.phase} ports=${snap.ports.length}`,
    )
  })
  await new Promise(() => {})
}
